### PR TITLE
Sample for demonstrating the use of AzureEventSourceListener (#23495)

### PR DESCRIPTION
**Pull request by andreyshihov**
# All SDK Contribution checklist:

This checklist is used to make sure that common guidelines for a pull request are followed.
- [x] **Please open PR in `Draft` mode if it is:**
	- Work in progress or not intended to be merged.
	- Encountering multiple pipeline failures and working on fixes.
- [ ] If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md).**
- [x] **The pull request does not introduce [breaking changes](https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/breaking-change-rules.md).**

### [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#general-guidelines)
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#testing-guidelines)
- [x] Pull request includes test coverage for the included changes.

### [SDK Generation Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#sdk-generation-guidelines)
- [ ] The generate.cmd file for the SDK has been updated with the version of AutoRest, as well as the commitid of your swagger spec or link to the swagger spec, used to generate the code. (Track 2 only)
- [ ] The `*.csproj` and `AssemblyInfo.cs` files have been updated with the new version of the SDK. Please double check nuget.org current release version.

## Additional management plane SDK specific contribution checklist: 
Note: Only applies to `Microsoft.Azure.Management.[RP]` or `Azure.ResourceManager.[RP]`
 
- [ ] Include updated [management metadata](https://github.com/Azure/azure-sdk-for-net/tree/main/eng/mgmt/mgmtmetadata).
- [ ] Update AzureRP.props to add/remove version info to maintain up to date API versions.

### Management plane SDK Troubleshooting
- If this is very first SDK for a services and you are adding new service folders directly under /SDK, please add `new service` label and/or contact assigned reviewer.
- If the check fails at the `Verify Code Generation` step, please ensure:
	- Do not modify any code in generated folders.
	- Do not selectively include/remove generated files in the PR.
	- Do use `generate.ps1/cmd` to generate this PR instead of calling `autorest` directly.
	Please pay attention to the @microsoft.csharp version output after running `generate.ps1`. If it is lower than current released version (2.3.82), please run it again as it should pull down the latest version.
	
	**Note: We have recently updated the PSH module called by `generate.ps1` to emit additional data. This would help reduce/eliminate the Code Verification check error. Please run following command**:

	    `dotnet msbuild eng/mgmt.proj /t:Util /p:UtilityName=InstallPsModules`

### Old outstanding PR cleanup
 Please note:
	If PRs (including draft) has been out for more than 60 days and there are no responses from our query or followups, they will be closed to maintain a concise list for our reviewers.

### CODE REVIEW

**Review by jsquire — COMMENTED** [review]
Thanks, @andreyshihov.   This looks really good!   I added some minor thoughts around wording and formatting, but the context and scope of the samples is spot on!

**Review by jsquire — APPROVED** [review]
Updates look great to me.   The only other outstanding task  that I can think of is to update the [README](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/eventhub/Azure.Messaging.EventHubs/samples/README.md) with the link to this sample.

We'll need to run the snippet generator to make CI happy:

`<repo root>\eng\scripts\Update-Snippets.ps1 -ServiceDirectory "eventhub"`

**Comment by andreyshihov** [comment]
All done.

Now I wonder, would it be helpful to include `Diagnostics.IsLoggingContentEnabled` example as demonstrated in [Azure SDK diagnostics](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core/samples/Diagnostics.md)

File: sdk/eventhub/Azure.Messaging.EventHubs/samples/README.md [hunk 1]
```diff
@@ -71,7 +71,10 @@ For the Event Hubs client library to interact with an Event Hub, it will need to
   A demonstration of how to interact with the client library using earlier versions of C#, where newer syntax for asynchronous enumeration and disposal are not available.
 
 - [Building a Custom Event Processor using EventProcessor&lt;TPartition&gt;](https://github.com/Azure/azure-sdk-for-net/tree/main/sdk/eventhub/Azure.Messaging.EventHubs/samples/Sample08_CustomEventProcessor.md)
```
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/samples/README.md:73]:
```suggestion
- [Building a Custom Event Processor using EventProcessor&lt;TPartition&gt;](https://github.com/Azure/azure-sdk-for-net/tree/main/sdk/eventhub/Azure.Messaging.EventHubs/samples/Sample08_CustomEventProcessor.md)  
```

File: sdk/eventhub/Azure.Messaging.EventHubs/samples/README.md [hunk 2]
```diff
@@ -71,7 +71,10 @@ For the Event Hubs client library to interact with an Event Hub, it will need to
...
-  An introduction to the `EventProcessor<TPartition>` base class which is used when building advanced processors which need full control over state management. 
+  An introduction to the `EventProcessor<TPartition>` base class which is used when building advanced processors which need full control over state management.
+
+- [Capturing Event Hubs logs using AzureEventSourceListener class](https://github.com/Azure/azure-sdk-for-net/tree/main/sdk/eventhub/Azure.Messaging.EventHubs/samples/Sample10_AzureEventSourceListener.md)
```
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/samples/README.md:76]:
```suggestion
- [Capturing Event Hubs logs using AzureEventSourceListener class](https://github.com/Azure/azure-sdk-for-net/tree/main/sdk/eventhub/Azure.Messaging.EventHubs/samples/Sample10_AzureEventSourceListener.md)  
```

File: sdk/eventhub/Azure.Messaging.EventHubs/samples/Sample10_AzureEventSourceListener.md [hunk 3]
```diff
@@ -0,0 +1,172 @@
+# Capturing Event Hubs logs using AzureEventSourceListener class
+
+The Event Hubs client library is instrumented using the .NET [`EventSource`](https://docs.microsoft.com/dotnet/api/system.diagnostics.tracing.eventsource) mechanism for logging. When instrumenting or diagnosing issues with applications that consume the library, it is often helpful to have access to the Event Hubs logs.  The following scenarios demonstrate how to use the [`AzureEventSourceListener`](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core/samples/Diagnostics.md#logging) from the `Azure.Core` package to capture logs emitted by the Event Hubs client library.
+
+## Capture ALL Events and display them in the Console
```
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/samples/Sample10_AzureEventSourceListener.md:5]:
```suggestion
## Capture ALL events and write them in to the console
```

File: sdk/eventhub/Azure.Messaging.EventHubs/samples/Sample10_AzureEventSourceListener.md [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+        throw new Exception($"The event could not be added.");
+    }
+}
+finally
+{
+    await producer.CloseAsync();
+}
+```
+
+## Capture ALL Events and display them in the Trace
```
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/samples/Sample10_AzureEventSourceListener.md:38]:
```suggestion
## Capture ALL events and write them to `Trace`
```

File: sdk/eventhub/Azure.Messaging.EventHubs/samples/Sample10_AzureEventSourceListener.md [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+        throw new Exception($"The event could not be added.");
+    }
+}
+finally
+{
+    await producer.CloseAsync();
+}
+```
+
+## Capture filtered Log Events
```
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/samples/Sample10_AzureEventSourceListener.md:67]:
```suggestion
## Apply filtering logic to logs
```

File: sdk/eventhub/Azure.Messaging.EventHubs/samples/Sample10_AzureEventSourceListener.md [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+        throw new Exception($"The event could not be added.");
+    }
+}
+finally
+{
+    await producer.CloseAsync();
+}
+```
+
+## Capture filtered Log Events to a file
```
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/samples/Sample10_AzureEventSourceListener.md:124]:
```suggestion
## Capture filtered logs to a file
```

File: sdk/eventhub/Azure.Messaging.EventHubs/samples/Sample10_AzureEventSourceListener.md [hunk 7]
```diff
@@ -0,0 +1,172 @@
+# Capturing Event Hubs logs using AzureEventSourceListener class
```
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/samples/Sample10_AzureEventSourceListener.md:1]:
```suggestion
# Capturing Event Hubs logs using the AzureEventSourceListener
```

**Review by jsquire — APPROVED** [review]
LGTM!

**Comment by jsquire** [comment]
Thank you, @andreyshihov, for all of your work on this.   The sample looks great, and I think it will be very helpful for the most common scenarios that we're asked about logging.   I made a couple of minor formatting and capitalization tweaks as a last pass.

Once the CI run completes,  I'll get this merged in!

**Comment by andreyshihov** [comment]
Thank you @jsquire, looking forward to contribute more... Do you know any issue I could have a look at considering my level of experience in this community?

**Comment by jsquire** [comment]
> Thank you @jsquire, looking forward to contribute more... Do you know any issue I could have a look at considering my level of experience in this community?

We always appreciate contributions.  😄    In general, [querying](https://github.com/Azure/azure-sdk-for-net/issues?q=is%3Aopen+is%3Aissue+label%3A%22help+wanted%22) for the "help wanted" tag is a good place to start.  Those will cover the gamut from "good-first-issue" to things that are complex and would require some deeper knowledge.  

For Event Hubs, if you're not bored by sample-related requests, there's "[Samples for Mocking Client Types](https://github.com/Azure/azure-sdk-for-net/issues/7382)" which is something that we see a lot of requests for and would be both highly visible and helpful to a wide audience.  There was a [draft PR](https://github.com/Azure/azure-sdk-for-net/pull/21630) that another community member had been working on, but there hasn't been any engagement for a few months and I fear has been abandoned.

There's also a [sample for integrating Event Hubs with ASP.NET](https://github.com/Azure/azure-sdk-for-net/issues/17683) which is another area that we see folks asking about.  This has a small bit of coverage in [this sample](https://github.com/Azure/azure-sdk-for-net/blob/main/samples/linecounter/Startup.cs) but is something that we could use some more Event Hubs-focused.   The integration is defined [here](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/eventhub/Azure.Messaging.EventHubs/src/Compatibility/EventHubClientBuilderExtensions.cs), and it would also be helpful to capture any gaps that we should address as part of this.

If you'd like to avoid samples, there's some test and infrastructure related items that would allow you to dig around in the codebase more and build some deeper familiarity:

- [Extend logging for better correlation and retry visibility](https://github.com/Azure/azure-sdk-for-net/issues/12566)
- [[Flaky Test] Event Hubs Function Bindings: Failures in UsGov and China Clouds](https://github.com/Azure/azure-sdk-for-net/issues/21300)
- [[Flaky Test] Event Hubs Function Bindings: EventHub_PartitionKey](https://github.com/Azure/azure-sdk-for-net/issues/23605)
- [[Flaky Test] Event Hubs Function Bindings: EventHub_InitialOffsetFromEnqueuedTime](https://github.com/Azure/azure-sdk-for-net/issues/21298)

If you're feeling like taking on something with more complexity, these are good candidates:

- [Sample: Tuning Guidance for Consuming Events](https://github.com/Azure/azure-sdk-for-net/issues/17681)
- [Event Processor Load Balancer: Allow for relinquishing a single partition](https://github.com/Azure/azure-sdk-for-net/issues/11078)
- [Revisit the ability to cache AMQP messages when adding events to the AmqpEventBatch](https://github.com/Azure/azure-sdk-for-net/issues/9569) _(this one is rough)_

**Comment by andreyshihov** [comment]
> For Event Hubs, if you're not bored by sample-related requests, there's "[Samples for Mocking Client Types](https://github.com/Azure/azure-sdk-for-net/issues/7382)" which is something that we see a lot of requests for and would be both highly visible and helpful to a wide audience. There was a [draft PR](https://github.com/Azure/azure-sdk-for-net/pull/21630) that another community member had been working on, but there hasn't been any engagement for a few months and I fear has been abandoned.

Hi @jsquire, I have just created draft PR #23677 to demonstrate use of Factory in the Handler tests. Am I moving in the right direction? Please share your thoughts. These are just first steps to understand if it conceptually makes sense, then I'll continue.

Thanks!

**Pull request merged**